code/sonalyze/sonarlog/persistentcluster.go

```diff
@@ -0,0 +1,553 @@
+	return d.sampleFiles
+}
+
+func (_ samplesAdapter) setFiles(d *persistentDir, files map[string]*LogFile) {
+	d.sampleFiles = files
+}
+
+func (_ samplesAdapter) proscribed(fn string) bool {
+	// Backward compatible: remove cpuhog.csv and bughunt.csv, these later moved into separate
+	// directory trees.
+	//
+	// TODO: REMOVEME: Remove once we know that all data stores have been cleaned of those files.
+	return fn == "cpuhog.csv" || fn == "bughunt.csv"
+}
+
+type sysinfoAdapter struct {
+}
+
+func (_ sysinfoAdapter) glob() string {
+	return "sysinfo-*.json"
+}
+
+func (_ sysinfoAdapter) contentAttribute() fileAttr {
+	return fileSonarSysinfo
+}
+
+func (_ sysinfoAdapter) getFiles(d *persistentDir) map[string]*LogFile {
+	return d.sysinfoFiles
+}
+
+func (_ sysinfoAdapter) setFiles(d *persistentDir, files map[string]*LogFile) {
+	d.sysinfoFiles = files
+}
+
+func (_ sysinfoAdapter) proscribed(fn string) bool {
+	return false
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+//
+// Low-level stuff.
+
+func (pc *PersistentCluster) findFilesLocked(
+	fromDate, toDate time.Time,
+	hosts *hostglob.HostGlobber,
+	fa filesAdapter,
+) ([]*LogFile, error) {
+	glob := fa.glob()
+	extensionLen := len(glob) - strings.LastIndexByte(glob, '.')
+
+	// Find all matching files in the date range.
+	fromDate = utime.ThisDay(fromDate)
+	toDate = utime.RoundupDay(toDate)
+
+	files := make([]*LogFile, 0)
+	for _, d := range pc.selectDirsLocked(fromDate, toDate) {
+		if fa.getFiles(d) == nil {
+			basenames := findFiles(pc.dataDir, d.name, glob)
+
+			// Filter file names that are simply disallowed
+			{
+				dest := 0
+				for src := 0; src < len(basenames); src++ {
+					fn := basenames[src]
+					if fa.proscribed(fn) {
+						continue
+					}
+					basenames[dest] = basenames[src]
+					dest++
+				}
+				basenames = basenames[:dest]
+			}
+
+			newFiles := make(map[string]*LogFile, len(basenames))
+			for _, name := range basenames {
+				f := newLogFile(
+					Fullname{
+						cluster:  pc.dataDir,
+						dirname:  d.name,
+						basename: name,
+					},
+					fileCacheable|fileAppendable|fa.contentAttribute(),
+				)
+				newFiles[name] = f
+			}
+			// This is wrong, it must be filtered *after*
+			fa.setFiles(d, newFiles)
+		}
+
+		// Retain only files whose names match the filter, if present
+		if hosts != nil && !hosts.IsEmpty() {
+			for _, c := range fa.getFiles(d) {
+				fn := c.basename
+				if hosts.Match(fn[:len(fn)-extensionLen]) {
+					files = append(files, c)
+				}
+			}
+		} else {
+			for _, c := range fa.getFiles(d) {
+				files = append(files, c)
+			}
+		}
+	}
+
+	return files, nil
+}
+
+// Pre: LOCK HELD
+// Pre: !ld.closed
+// Post if !error: directory exits on disk
+// Post if !error: directory exists in cluster's tree
+func (pc *PersistentCluster) findFileByTimeLocked(timestamp, filename string, fa filesAdapter) (*LogFile, error) {
+	tval, err := time.Parse(time.RFC3339, timestamp)
```

Comment:
UTC or not?  And how is this handled deeper down?  ensureScannedDirectoryLocked cleans up from/to, but the cleaned times are not presented to the binary search, for example.

Reply:
Cleaned this up.  We could do even better by defining a local UtcTime type, I think, and encapsulate this invariant.